### DISCUSSION TITLE
HTTP String und JSON Pfad beim Shelly PRO 3 em

### DISCUSSION THREAD

**Discussion by OlSteph** [hoylabs/OpenDTU-OnBattery#247] [Q&A] [answered]
Hallo,

ich habe mir einen Shelly Pro 3em besorgt und würde diesen auch hier gerne verbinden.
Aktuell ist der Powerfox/Poweropti angebunden - die 10 Sekündige Abfrage über die Cloud würde ich gerne ersetzen.

Via http://192.168.179.128/rpc/Shelly.GetStatus?id=0 bekomme ich nach Eingabe von Benutzername (admin) und Passwort (xxxx)
folgende Seite angezeigt

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/00e1867a-3cfe-4c60-b16d-ac5261120a4b)

Als JSON Path vermute ich hier: total_act_power

Mir ist hier (bin da immer noch Laie auf dem Gebiet) nicht klar, wie ich das bei den Einstellungen eingeben kann.
Versuchsweise in Anlehnung an die Powerfox Konfig (ohne Base64 Codierte ANmeldedaten) waren erfolglos.
Danke und Gruß

**Comment by helgeerbe** [comment]
Ich kenne zwar den pro 3em nicht, aber wenn ich https://shelly-api-docs.shelly.cloud/gen2/ComponentsAndServices/EM richtig verstanden habe, würde ich folgendes mal versuchen:  `http://admin:PASSWORD@192.168.179.128/rpc/EM.GetStatus?id=0`

Als JSON Path: total_act_power

**Reply by OlSteph** [reply]
Bad HTTP code: 401 ist die Meldung.
Damit hatte ich auch schon gespielt, aber es nicht hinbekommen.
Habe mir diverse Seiten von Shelly dazu angeschaut - am Besten passte aus meiner Sicht folgende:
https://shelly-api-docs.shelly.cloud/gen2/General/Authentication
-> aber hieraus wurde ich auch nicht schlau, wie ich das in der DTU umsetzen kann/muss

**Comment by helgeerbe** [comment]
Oh, der Shelly Pro 3em verwendet digest authentication anstelle von basic authentication. Ich fürchte, das ist nicht implementiert.

Gibt es Freiwillige, die Zeit haben das zu implementieren?

**Reply by OlSteph** [reply]
Hallo,

vielleicht bin ich auf dem Holzweg - vielleicht fehlt mir aber auch das Verständnis - ich bin zumindest bis zur Success Meldung bekommen - aber schaffe es nicht den gewünschten Wert aus total_act_power zu bekommen - vielleicht muss ich de Pfad hier anders angeben?

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/7ce4b3b1-3565-421f-85e2-825d56ca77f3)

 wenn ich beim Pfad total_act_power eingebe kommt:
 
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/576710b4-3b25-4a43-b541-3023fd32181e)

Den Inhalt aus Optional: "HTTP request header - Wert" habe ich von der Seite http://192.168.179.128/rpc/Shelly.GetStatus unter Kopfzeilen manuell kopiert (nach erfolgreicher Anmeldung mit Benutzername und Kennwort)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/dbe80925-d784-4fcc-a7b0-7a8023b69b77)

Ist das ggf ein Ansatz oder eine falsche Erfolgsmeldung des Tests?

**Reply by spcqike** [reply]
Probier mal 

em/0/total_act_power

**Reply by helgeerbe** [reply]
Wäre ein Ansatz. Kann funktionieren, wenn der Shelly den Zufallswert tatsächlich nicht ändert. Ich habe jetzt nicht in die Spec geschaut. Aber eigentlich sollten die Werte geändert werden, um Replay Attacken zu unterbinden.

**Reply by MalteSchm** [reply]
Shelly 3EM
URL: http://deine URL/status
Jsonpath: total_power

So läuft der jedenfalls bei mir. Bei Eingabe der URL im Browser solltest du schon Daten sehen

EDIT: Ah vergiss es. Ich habe übersehen das du den PRO hast

**Reply by OlSteph** [reply]
Nach erneutem Test verändern dich die "Zahlen" nach neuer Anmeldung über die Weboberfläche.
Der Test mittels em/o/total_act_power ergibt ebenfalls einen Fehler:

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/1cc42e53-f307-41ae-9625-ae43df865fe7)

**Reply by spcqike** [reply]
Und em:0/total_act_power? 

Auch wenn „em:0“ komisch aussieht und ich mir nicht sicher bin, dass das ein gültiger json so macht. 

Wie sehen die rohdaten des json aus?

**Reply by OlSteph** [reply]
Hier die Rohdaten:
{
  "ble": {},
  "cloud": {
    "connected": true
  },
  "em:0": {
    "id": 0,
    "a_current": 0.474,
    "a_voltage": 233.1,
    "a_act_power": 35.9,
    "a_aprt_power": 110.4,
    "a_pf": -0.6,
    "b_current": 0.443,
    "b_voltage": 233.2,
    "b_act_power": 25.8,
    "b_aprt_power": 103.2,
    "b_pf": -0.58,
    "c_current": 0.742,
    "c_voltage": 233.9,
    "c_act_power": 98.8,
    "c_aprt_power": 173.3,
    "c_pf": -0.71,
    "n_current": null,
    "total_current": 1.658,
    "total_act_power": 160.488,
    "total_aprt_power": 386.886,
    "user_calibrated_phase": []
  },
  "emdata:0": {
    "id": 0,
    "a_total_act_energy": 4663.82,
    "a_total_act_ret_energy": 6028.1,
    "b_total_act_energy": 4175.87,
    "b_total_act_ret_energy": 10465.64,
    "c_total_act_energy": 10398.16,
    "c_total_act_ret_energy": 0,
    "total_act": 19237.85,
    "total_act_ret": 16493.74
  },
  "eth": {
    "ip": null
  },
  "modbus": {},
  "mqtt": {
    "connected": true
  },
.....

**Reply by OlSteph** [reply]
Beim vorgeschlagenen Wert, auch mal mit und ohne vorangehendes und abschließendes "/"
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/1e1633f0-523e-49b2-8837-9c5f4ea85c9a)
-> es bleibt auf dieser Meldung stehen

**Reply by OlSteph** [reply]
Zum Querprüfen hab ich das Gerät mal ins LAN gehängt (daher andere IP) und die Authentifizierung abgeschaltet (Zugangskennwort deaktiviert)

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/92075d33-a856-4f2b-a37a-f4833acb2dbf)

**Comment by OlSteph** [comment]
Eine andere Möglichkeit wäre MQTT - hier stoße ich aber noch an meine Grenzen, da die Objekte im IOBroker eine Art Array (JSON oder sowas) sind und nicht eigene Datenpunkte.

Ggf könnte mir nier jemand helfen, wie dies als MQTT Anbindung gelöst werden könnte? Dann umgehen wir auch ggf das Problem der Digest-Authentifizierung.
(Kopierter Pfad aus IOBroker)
mqtt.0.shellypro3em-ecxxxxxxxxx8.status.em:0 
Inhalt des Objektes em:0
{"id":0,"a_current":0.500,"a_voltage":232.7,"a_act_power":104.2,"a_aprt_power":116.2,"a_pf":-0.90,"b_current":0.449,"b_voltage":232.9,"b_act_power":28.1,"b_aprt_power":104.4,"b_pf":-0.58,"c_current":0.687,"c_voltage":233.5,"c_act_power":92.6,"c_aprt_power":160.0,"c_pf":-0.70,"n_current":null,"total_current":1.635,"total_act_power":224.834,"total_aprt_power":380.679, "user_calibrated_phase":[]}

Wird gespeichert - aber Anzeige bleibt auf "0"
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/2bacef59-4907-42ca-b490-043683150958)

Objektdaten:
{
  "_id": "mqtt.0.shellypro3em-ecxxxxxxxxx8.status.em:0",
  "common": {
    "name": "shellypro3em-ecxxxxxxxxx8/status/em:0",
    "write": true,
    "read": true,
    "role": "variable",
    "desc": "mqtt server variable",
    "type": "string"
  },
  "native": {
    "topic": "shellypro3em-ecxxxxxxxxx/status/em:0"
  },
  "type": "state",
  "from": "system.adapter.admin.0",
  "user": "system.user.admin",
  "ts": 1685624662617
}

**Comment by helgeerbe** [comment]
I just created a feature request #251 to support digest authentication.

**Comment by helgeerbe** [comment]
Should work with release 2023.07.04. Please provide feedback, because I don't own such a device.

**Reply by OlSteph** [reply]
Hello @helgeerbe - i will test it if it is available as image. Thank you (and ChatGPT) :-)

**Reply by helgeerbe** [reply, marked as answer]
Thanks, version should be built by now.  ChatGPT was not so helpful at the end. I had to read the specs and fix the code. ChatGPT was not able to generate working code, even when you point out what is wrong.

**Reply by OlSteph** [reply]
@helgeerbe (ChatGPT is nice, but yet far away from real knowledge - i think)

I flashed the new version and tried to connect to my Shelly pro3em:
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/ced44b03-998a-43de-aa1c-1a2ed9661077)
Looks very good :-)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/c15a5142-9676-487b-8796-a6cc4655b67e)

Thank you very much - with that digest implementation this should work for all Shelly Gen.2 devices

**Reply by helgeerbe** [reply]
Thanks for the feedback. Happy that it works.

**Comment by gruenteefan** [comment]
Hallo zusammen, 
ich bin auf der Suche nach der korrekten Eingabe der Daten zur Abfrage der Verbrauchsdaten aus meinem Shelly Pro 3EM in OpenDTU / Stromzähler. Die Freigabe/ Anbindung des Shellys im MQTT-Broker ist bereits erfolgt. Ich schaffe es nur nicht, die Eingaben in den Einstellungen des OpenDTUs zu bewerkstelligen. Kann mir hier jemand helfen?

**Reply by Manos1966** [reply]
Hallo @gruenteefan 

Vielleicht hier:

https://github.com/hoylabs/OpenDTU-OnBattery/wiki/Power-Meter

https://github.com/hoylabs/OpenDTU-OnBattery/issues/364

**Comment by gruenteefan** [comment]
An dieser Stelle einmal eine grundsätzliche Frage: Ich versorge die DTU gerade mit den Shelly-Pro-3EM-Daten mit den HTTP(S) - JSON-Daten. Wäre es besser/ effizienter, hier die MQTT-Daten zu verwenden? Was davon geht schneller/ liefert die Daten schneller an die DTU?

**Reply by schlimmchen** [reply]
Das kann ich dir pauschal nur teilweise beantworten:

MQTT ist definitiv effizienter.

MQTT ist aber nur dann schneller, wenn dein Shelly die Messdaten auch zügig (normalerweise im Sekundentakt) über MQTT rausbläst. Tut er das? Mein Shelly 3EM hat Tasmota und der tut es mit Hilfe einer passenden Rule (nur Tasmota):

```
ON energy#total DO publish grid/energy/total %value% ENDON ON energy#power[1] DO publish grid/power/phase1/watts %value% ENDON ON energy#power[2] DO publish grid/power/phase2/watts %value% ENDON ON energy#power[3] DO publish grid/power/phase3/watts %value% ENDON
```

**Comment by gruenteefan** [comment]
So wie ich das sehe, werden die Leistungsdaten in MQTT alle in einer Zeile veröffentlich. Ich kann also so einfach kein Topic für Wert 1, 2 und 3 eintragen. Kann mir da jemand helfen?

**Reply by schlimmchen** [reply]
> alle in einer Zeile veröffentlich

Was heißt das, bitte? Unsere MQTT PowerMeter Implementierung kann auch JSON auseinander nehmen.

**Comment by gruenteefan** [comment]
Sorry - was würde ich denn bei den MQQT-Topics 1, 2 und 3 genau eintragen?

**Comment by gruenteefan** [comment]
Wäre das dann dieser Pfad hier?

Hier werde die Werte auf keinen Fall sekündlich aktualisiert - das dauert schon viel länger (5-6 Sekunden). Wie kann ich hier erreichen, dass eine sekündliche Aktualisierung klappt?

![Screenshot_MQTT_PowerMeter](https://github.com/user-attachments/assets/da1ee036-26d5-434d-95e6-6d923aef068e)

**Reply by schlimmchen** [reply]
Danke für den Screenshot. Also wie du deinen Shelly dazu bringst, häufiger was zu veröffentlichen weiß ich nicht. Mach Tasmota drauf, dann kann ich Tips geben :wink:

Die Topics die du eingerahmt hast, sind jene, die OpenDTU-OnBattery ermittelt und selbst veröffentlicht hat. Die kannst du also nicht wieder einlesen, das macht keinen Sinn. Die sind dort, weil OpenDTU-OnBattery die PowerMeter Werte aus Quellen anders als MQTT selbst auch in MQTT veröffentlicht.

Und wie du an die Daten des markierten Topics kommst, kann ich dir sagen, wenn du das mal als Text vollständig als Kommentar teilst.

**Comment by gruenteefan** [comment]
meinst Du das hier?

```json
{
  "src": "shellypro3em-ec6260963964",
  "dst": "shellypro3em-ec6260963964/events",
  "method": "NotifyStatus",
  "params": {
    "ts": 1738960366.89,
    "em:0": {
      "id": 0,
      "a_act_power": 214.2,
      "a_aprt_power": 251,
      "a_current": 1.096,
      "a_freq": 50,
      "a_pf": 0.85,
      "a_voltage": 229.1,
      "b_act_power": -179.7,
      "b_aprt_power": 215.2,
      "b_current": 0.936,
      "b_freq": 50,
      "b_pf": 0.83,
      "b_voltage": 229.8,
      "c_act_power": 57.3,
      "c_aprt_power": 80.3,
      "c_current": 0.35,
      "c_freq": 50,
      "c_pf": 0.72,
      "c_voltage": 229.2,
      "n_current": null,
      "total_act_power": 91.801,
      "total_aprt_power": 546.482,
      "total_current": 2.383
    }
  }
}
```

**Reply by schlimmchen** [reply]
Der JSON Pfad ist dann "params/em:0/total_act_power".

**Comment by gruenteefan** [comment]
Prima. Den schreib ich dann in das Feld "Optional: JSON-Pfad" bei "Konfiguration Wert 1" rein, oder? und als TOPIC dann 
shellypro3em-ec6260963964/events/rpc

Ist das soweit korrekt?

**Reply by schlimmchen** [reply]
Ich denke schon. Es wird jedenfalls nichts kaputt gehen, also erstmal ausprobieren :wink:
